torch/cuda/__init__.py

```diff
@@ -317,41 +318,93 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
+
+    def __init__(self, stream: Optional['torch.classes.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        self.prev_stream_device_index = -1
+        self.cur_stream_device_index = -1
+        # Initialize the below streams to default stream on the current device
+        self.device_index = _get_device_index(device=None, optional=True)
+        self.src_prev_stream = torch.cuda.default_stream(self.device_index)
+        self.dst_prev_stream = torch.cuda.default_stream(self.device_index)
 
-    if src_prev_stream.device != stream.device:
-        # The given stream is on a different device; have to restore the
-        # current_stream on that device on exit as well
-        with device(stream.device):
-            dst_prev_stream = current_stream()
+    def __enter__(self):
+        self.idx = _get_device_index(device=None, optional=True)
+        # If there is no CUDA device available, return
+        if self.idx == -1:
+            return
 
-    torch._C._cuda_setStream(stream._cdata)
-    try:
-        yield
-    finally:
-        if src_prev_stream.device != stream.device:
-            torch._C._cuda_setStream(dst_prev_stream._cdata)
-        torch._C._cuda_setStream(src_prev_stream._cdata)
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
+        # Return if stream is None
+        if cur_stream is None:
+            return
+        self.src_prev_stream = torch.cuda.current_stream(self.idx)
+
+        # Eager mode and script mode APIs to access the stream device and the device index
+        # vary. Hence, we have to check the mode we are in and call the respective API.
+        #
+        # Note: The API's don't differ in functionality, only differ at the API level.
+        # In Eager Mode, the stream object doesn't have a function to access device index,
+        # hence we use `_get_device_index`. In script mode, device_index is easily
+        # accessible by calling stream.device_index(). With device, in Eager mode to access
+        # device of the stream can be accessed by stream.device while in JIT we need to use
+        # stream.device(). But both return the device object of type `torch.device`
+        if torch.jit.is_scripting():
+            self.prev_stream_device_index = self.src_prev_stream.device_index()  # type: ignore
+            self.cur_stream_device_index = cur_stream.device_index()
+            cur_stream_device = cur_stream.device()
+        else:
+            self.prev_stream_device_index = _get_device_index(self.src_prev_stream.device, optional=True)
+            self.cur_stream_device_index = _get_device_index(cur_stream.device, optional=True)
+            cur_stream_device = cur_stream.device
+
+        # If the stream is not on the current device, then change the device
+        # and set the current stream on the device
+        if self.prev_stream_device_index != self.cur_stream_device_index:
```

Comment:
In the old implementation, this line was
```
if src_prev_stream.device != stream.device
```
Can we expose the device of a stream as a property using #50670 and then avoid storing device index explicitly like this?

Reply:
Torchscript doesn't support comparing two `device` objects. So even if we expose, it won't be of any help.

Reply:
Can we add a JIT operator for it? There is a an `operator==` defined for `c10::Device`.